django/thunderstore/api/cyberstorm/views/package_preview.py

```diff
@@ -0,0 +1,288 @@
+
+        params: OrderedDict = qp.validated_data
+        listings_qs = self.get_queryset(params)
+        listings_qs = self.order_queryset(params["ordering"], listings_qs)
+        categories = PackageCategory.objects.filter(
+            pk__in=listings_qs.aggregate(arr=ArrayAgg("categories", distinct=True))[
+                "arr"
+            ]
+        )
+        listings_page = self.paginate(params, listings_qs, params["page_size"])
+        listing_count = len(listings_qs)
+        serializer = self.serialize_results(
+            listings_page, categories, listing_count, params["page_size"]
+        )
+
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+    def get_full_cache_vary(self, params: OrderedDict) -> str:
+        cache_vary = "package_cache_key"
+        if params.get("community_identifier", None):
+            cache_vary += f".{params['community_identifier']}"
+        if params.get("namespace_identifier", None):
+            cache_vary += f".{params['namespace_identifier']}"
+        if params.get("team_identifier", None):
+            cache_vary += f".{params['team_identifier']}"
+        if params.get("user_identifier", None):
+            cache_vary += f".{params['user_identifier']}"
+        if params.get("package_identifier", None):
+            cache_vary += f".{params['package_identifier']}"
+        cache_vary += f".{params['deprecated']}"
+        cache_vary += f".{params['nsfw']}"
+        cache_vary += f".{sorted(params['included_categories'])}"
+        cache_vary += f".{sorted(params['excluded_categories'])}"
+        cache_vary += f".{params.get('section', '-')}"
+        cache_vary += f".{params.get('q', '-')}"
+        cache_vary += f".{params['ordering']}"
+        cache_vary += f".{params['page']}"
+        cache_vary += f".{params['page_size']}"
+        return cache_vary
+
+    def get_queryset(self, params) -> QuerySet[PackageListing]:
+        # Prep Q objects
+        filter_q = Q()
+        exclude_q = Q()
+
+        # Add majority of filters
+        if params.get("community_identifier", None):
+            filter_q.add(Q(community__identifier=params["community_identifier"]), Q.AND)
+
+        if params.get("namespace_identifier", None):
+            filter_q.add(
+                Q(package__namespace__name=params["namespace_identifier"]), Q.AND
+            )
+
+        if params.get("team_identifier", None):
+            filter_q.add(Q(package__owner__name=params["team_identifier"]), Q.AND)
+
+        if params.get("user_identifier", None):
+            filter_q.add(
+                Q(
+                    package__owner__members__user__username__icontains=params[
+                        "user_identifier"
+                    ]
+                ),
+                Q.AND,
+            )
+
+        if params.get("package_identifier", None):
+            filter_q.add(Q(package__name=params["package_identifier"]), Q.AND)
```

Comment:
I've made something, but it's not pretty, since I can't think straight with the headache I've got. Let me know if you had something cleaner in mind 👍